core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ipv4/IPv4RangeBuilder.java

```diff
@@ -110,57 +115,75 @@ public IPv4RangeBuilder addUnboundedFrom(String from) {
         }
-        int addr = (( Integer.parseInt(parts[0]) << 24 ) & 0xFF000000)
-                | (( Integer.parseInt(parts[1]) << 16 ) & 0xFF0000)
-                | (( Integer.parseInt(parts[2]) << 8 ) & 0xFF00)
-                |  ( Integer.parseInt(parts[3]) & 0xFF);
-
-        int mask = (-1) << (32 - Integer.parseInt(parts[4]));
-
-        if (Integer.parseInt(parts[4]) == 0) {
-            mask = 0 << 32;
+        String[] ipAddressFields = fields[0].split("\\.");
+        if (ipAddressFields.length != 4) {
+            throw new IllegalArgumentException(
+                    String.format(Locale.ROOT, "expected [a, b, c, d] but was [%s] after splitting on \".\" before \"/\" in [%s]", Arrays.toString(ipAddressFields), cidr)
+            );
         }
-
-        int from = addr & mask;
-        long longFrom = intIpToLongIp(from);
-        if (longFrom == 0) {
-            longFrom = -1;
+        long accumulator = 0;
+        for (int i = 0; i < 4; i++) {
+            int octet;
+            try {
+                octet = Integer.parseInt(ipAddressFields[i]);
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException(
+                        String.format(Locale.ROOT, "invalid octet [%s] in position [%d] in [%s]", ipAddressFields[i], i, cidr),
+                        e
+                );
+            }
+            if (octet < 0 || octet > 255) {
+                throw new IllegalArgumentException(
+                        String.format(Locale.ROOT, "invalid octet [%s], out of range in position [%d] in [%s]", ipAddressFields[i], i, cidr)
+                );
+            }
+            accumulator <<= 8;
+            accumulator |= octet;
         }
 
-        int to = from + (~mask);
-        long longTo = intIpToLongIp(to) + 1; // we have to +1 here as the range is non-inclusive on the "to" side
-
-        if (longTo == InternalIPv4Range.MAX_IP) {
-            longTo = -1;
+        int networkMask;
+        try {
+            networkMask = Integer.parseInt(fields[1]);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException(
+                    String.format(Locale.ROOT, "invalid network mask [%s] in [%s]", fields[1], cidr),
+                    e
+            );
+        }
+        if (networkMask < 0 || networkMask > 32) {
+            throw new IllegalArgumentException(
+                    String.format(Locale.ROOT, "invalid network mask [%s], out of range in [%s]", fields[1], cidr)
+            );
         }
 
-        return new long[] { longFrom, longTo };
-    }
+        long blockSize = 1L << (32 - networkMask);
+        // validation
+        if ((accumulator & (blockSize - 1)) != 0) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            Locale.ROOT,
+                            "invalid address/network mask combination in [%s]; only the most significant [%d] bits can be set",
```

Comment:
I'm thinking it could be more user-friendly to suggest a correction, like `"did you mean " + ipToString(accumulator & (blockSize - 1)) + "/" + networkMask` in the error message